src/coroutine.rs

```diff
@@ -0,0 +1,218 @@
+use crate::{intern, IntoPy, Py, PyAny, PyErr, PyObject, PyResult, Python};
+use futures::task::ArcWake;
+use pyo3_macros::{pyclass, pymethods};
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+const COROUTINE_REUSED_ERROR: &str = "cannot reuse already awaited coroutine";
+
+/// Python coroutine wrapping a [`Future`].
+#[pyclass(crate = "crate")]
+pub struct Coroutine {
+    future: Option<Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>>,
+    #[allow(clippy::type_complexity)]
+    throw_callback: Option<Box<dyn FnMut(&PyAny) + Send>>,
+    waker: Option<Arc<PyFutureWaker>>,
+}
+
+impl Coroutine {
+    /// Wrap a boxed future in to a Python coroutine.
+    ///
+    /// Coroutine `send` polls the wrapped future, ignoring the value passed
+    /// (should always be `None` anyway).
+    ///
+    /// If `throw_callback` has been provided, coroutine `throw` calls it
+    /// with the exception passed, and poll the wrapped future.
+    /// Otherwise, the exception is reraised and the future dropped.
+    #[allow(clippy::type_complexity)]
+    pub fn new(
+        future: Pin<Box<dyn Future<Output = PyResult<PyObject>> + Send>>,
+        throw_callback: Option<Box<dyn FnMut(&PyAny) + Send>>,
+    ) -> Self {
+        Self {
+            future: Some(future),
+            throw_callback,
+            waker: None,
+        }
+    }
+
+    ///  Wrap a generic future into a Python coroutine.
+    pub fn from_future<F, T, E>(future: F) -> Self
+    where
+        F: Future<Output = Result<T, E>> + Send + 'static,
+        T: IntoPy<PyObject> + Send,
+        E: Send,
+        PyErr: From<E>,
+    {
+        let wrap = async move {
+            let obj = future.await?;
+            // SAFETY: GIL is acquired when future is polled (see `Coroutine::poll`)
+            Ok(obj.into_py(unsafe { Python::assume_gil_acquired() }))
+        };
+        Self::new(Box::pin(wrap), None)
+    }
+
+    fn poll(
+        &mut self,
+        py: Python<'_>,
+        throw: Option<&PyAny>,
+    ) -> PyResult<IterNextOutput<PyObject, PyObject>> {
+        // raise if the coroutine has already been run to completion
+        let future_rs = match self.future {
+            Some(ref mut fut) => fut,
+            None => return Err(PyRuntimeError::new_err(COROUTINE_REUSED_ERROR)),
+        };
+        // pass thrown exception to throw callback if there is one, otherwise reraise it
+        if let Some(exc) = throw {
+            if let Some(cb) = &mut self.throw_callback {
+                cb(exc)
+            } else {
+                self.future.take();
+                return Err(PyErr::from_value(exc));
+            };
+        }
+        loop {
+            // create a waker from an `asyncio.Future`, or try to modify it in place
+            if let Some(waker) = self.waker.as_mut().and_then(Arc::get_mut) {
+                waker.future = asyncio_future(py)?;
```

Comment:
So if I read it correctly, with this PR the way that we support additional async runtimes on the Python side would be by making the `waker` able to support various async interfaces?

Reply:
Exactly. That's what I've done in *pyo3-async*, with a [`CoroutineWaker`](https://github.com/wyfo/pyo3-async/blob/e55eb3b5d8743693ceee1ffa99689fb3e64b7ba0/src/coroutine.rs#L15) trait and its implementations for *asyncio*, *trio* and *anyio*.